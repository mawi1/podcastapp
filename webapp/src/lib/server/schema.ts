import {
  boolean,
  customType,
  doublePrecision,
  integer,
  pgTable,
  primaryKey,
  text,
  timestamp,
  unique,
} from "drizzle-orm/pg-core";

export const user = pgTable("user_", {
  id: integer("id").primaryKey().generatedAlwaysAsIdentity(),
  username: text("username").unique("user__username__unique").notNull(),
  passwordHash: text("password_hash").notNull(),
});

export const session = pgTable("session", {
  id: text("id").primaryKey(),
  userId: integer("user_id")
    .notNull()
    .references(() => user.id, { onDelete: "cascade" }),
  expiresAt: timestamp("expires_at", {
    withTimezone: true,
    mode: "date",
  }).notNull(),
});

const bytea = customType<{ data: Uint8Array }>({
  dataType() {
    return "bytea";
  },
});

export const podcast = pgTable("podcast", {
  id: integer("id").primaryKey().generatedAlwaysAsIdentity(),
  title: text("title").notNull(),
  link: text("link").notNull(),
  description: text("description").notNull(),
  feedUrl: text("feed_url").notNull().unique("podcast__feed_url__unique"),
  updated: boolean("updated").notNull().default(true),
  httpEtag: bytea("http_etag"),
});

export const episode = pgTable(
  "episode",
  {
    id: integer("id").primaryKey().generatedAlwaysAsIdentity(),
    title: text("title"),
    content: text("content"),
    audioUrl: text("audio_url"),
    pubDate: timestamp("pub_date", {
      withTimezone: true,
      mode: "date",
    }).notNull(),
    durationSeconds: integer("duration_seconds"),
    guid: text("guid").notNull(),
    podcastId: integer("podcast_id")
      .notNull()
      .references(() => podcast.id, { onDelete: "cascade" }),
  },
  (t) => [unique("episode__guid__podcast_id__unique").on(t.guid, t.podcastId)]
);

export const subscription = pgTable(
  "subscription",
  {
    userId: integer("user_id")
      .notNull()
      .references(() => user.id, { onDelete: "cascade" }),
    podcastId: integer("podcast_id")
      .notNull()
      .references(() => podcast.id, { onDelete: "cascade" }),
  },
  (t) => [primaryKey({ columns: [t.userId, t.podcastId] })]
);

export const playlist = pgTable(
  "playlist",
  {
    userId: integer("user_id")
      .notNull()
      .references(() => user.id, { onDelete: "cascade" }),
    episodeId: integer("episode_id")
      .notNull()
      .references(() => episode.id, { onDelete: "cascade" }),
    position: integer("position").notNull(),
  },
  (t) => [
    primaryKey({ columns: [t.userId, t.episodeId] }),
    unique("playlist__user_id__position__unique").on(t.userId, t.position),
  ]
);

export const resumeData = pgTable(
  "resume_data",
  {
    userId: integer("user_id")
      .notNull()
      .references(() => user.id, { onDelete: "cascade" }),
    episodeId: integer("episode_id")
      .notNull()
      .references(() => episode.id, { onDelete: "cascade" }),
    currentTime: doublePrecision("current_time").notNull(),
    playbackRate: doublePrecision("playback_rate").notNull(),
  },
  (t) => [primaryKey({ columns: [t.userId, t.episodeId] })]
);
