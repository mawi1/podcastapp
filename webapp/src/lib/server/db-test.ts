import { type NodePgDatabase, drizzle } from "drizzle-orm/node-postgres";
import { migrate } from "drizzle-orm/postgres-js/migrator";
import path, { dirname } from "path";
import { Client, Pool } from "pg";
import sha1 from "simple-sha1";
import { fileURLToPath } from "url";
import { expect, inject, test, vi } from "vitest";

import * as schema from "./schema";
import { episode, podcast } from "./schema";

export type DB = NodePgDatabase<typeof import("./schema")> & { $client: Pool };

export async function insertTestUser(db: DB): Promise<void> {
  const passwordHash =
    "$argon2id$v=19$m=19456,t=2,p=1$C4ma4F2Ing8dghbBKekbmw$L6K7xP4lGY2jI79nheDwCK1QXYiKShcJtOJ4t4xdRBg"; // password111
  await db.insert(schema.user).values({ username: "testuser", passwordHash });
}

export async function insertTestPodcasts(db: DB) {
  const insert = async (title: string): Promise<number> => {
    const link = `http://${title.replaceAll(" ", "").toLocaleLowerCase()}.com`;
    const feedUrl = link + "/feed.xml";
    const description = `Podcast ${title} description`;
    const podcastId = (
      await db
        .insert(podcast)
        .values({ title: title, link, description, feedUrl })
        .returning({ id: podcast.id })
    )[0].id;
    return podcastId;
  };

  const podcasts = [
    {
      title: "The Startup Squad",
      episodes: 34,
      // id 1
    },
    {
      title: "Love in the Algorithm",
      episodes: 101,
      // id 2
    },
    {
      title: "Murder at the Microphone",
      episodes: 45,
      // id 3
    },
    {
      title: "The Paranormal Pundits",
      episodes: 29,
      // id 4
    },
    {
      title: "Tech Tales from the Future",
      episodes: 255,
      // id 5
    },
    {
      title: "Fitness Frontier",
      episodes: 12,
      // id 6
    },
    {
      title: "Culinary Crime Scene",
      episodes: 22,
      // id 7
    },
    {
      title: "The Historical Haunts",
      episodes: 612,
      // id 8
    },
    {
      title: "8-Bit Adventures",
      episodes: 93,
      // id 9
    },
    {
      title: "12 Steps to Success",
      episodes: 278,
      // id 10
    },
    {
      title: "@Tech Talk ",
      episodes: 145,
      // id 11
    },
    {
      title: "+Positive Vibes",
      episodes: 102,
      // id 12
    },
  ];
  const startDate = new Date(Date.UTC(2024, 1, 1, 0, 0, 0, 0)).getTime();
  for (const p of podcasts) {
    const podcastId = await insert(p.title);
    for (let i = 1; i <= p.episodes; i++) {
      const title = `Podcast (${podcastId}) Episode #${i}`;
      const content = "<p>Content ${title}</p>";
      const audioUrl = `http://podcastapp.mawi.cc/podcast/${podcastId}/${i}/audio.mp3`;
      const guid = `${podcastId}_${i}`;
      const durationSeconds = 234 + 3 * i;
      const pubDate = new Date(startDate + podcastId * 8 + i * 3600000 * 17);
      await db
        .insert(episode)
        .values({ podcastId, title, content, audioUrl, guid, durationSeconds, pubDate });
    }
  }
}

interface Fixture {
  db: DB;
}

type PostgresConfig = {
  user: string;
  password: string;
  host: string;
  port: number;
};

async function run_sql(config: PostgresConfig, sql: string) {
  const client = new Client(config);
  await client.connect();
  client.on("error", (err) => {
    console.error("something bad has happened!", err.stack);
  });
  await client.query(sql);
  await client.end();
}

declare module "vitest" {
  export interface ProvidedContext {
    postgresConfig: {
      user: string;
      password: string;
      host: string;
      port: number;
    };
  }
}

export const dbTest = test.extend<Fixture>({
  // eslint-disable-next-line no-empty-pattern
  db: async ({}, use) => {
    vi.resetModules();

    const config = inject("postgresConfig");

    if (expect.getState().currentTestName === undefined) {
      throw "no test name provided";
    }
    const dbName = "D" + sha1.sync(expect.getState().currentTestName!);
    await run_sql(config, `CREATE DATABASE "${dbName}";`);

    const migrationsFolder = path.resolve(
      dirname(fileURLToPath(import.meta.url)),
      "..",
      "..",
      "..",
      "drizzle"
    );
    const pool = new Pool({
      user: config.user,
      password: config.password,
      host: config.host,
      port: config.port,
      database: dbName,
    });
    const db = drizzle(pool, { schema });
    await migrate(db, { migrationsFolder });

    vi.doMock("./db.ts", () => {
      return {
        default: db,
      };
    });
    await use(db);
    await pool.end();
    await run_sql(config, `DROP DATABASE "${dbName}";`);
  },
});
