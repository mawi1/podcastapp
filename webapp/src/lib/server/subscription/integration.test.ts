import { and, count, eq } from "drizzle-orm";
import { describe, expect } from "vitest";

import { type DB, dbTest, insertTestPodcasts, insertTestUser } from "../db-test";
import { subscription } from "../schema";

async function countRows(db: DB, userId: number, podcastId: number): Promise<number> {
  return (
    await db
      .select({ count: count() })
      .from(subscription)
      .where(and(eq(subscription.userId, userId), eq(subscription.podcastId, podcastId)))
  )[0].count;
}

describe("playlist", () => {
  dbTest("subscribe", async ({ db }) => {
    const { subscribe } = await import(".");

    await insertTestUser(db);
    await insertTestPodcasts(db);

    await subscribe(1, 7);
    const rows = await countRows(db, 1, 7);
    expect(rows).toBe(1);
  });

  dbTest("subscribe is idempotent", async ({ db }) => {
    const { subscribe } = await import(".");

    await insertTestUser(db);
    await insertTestPodcasts(db);

    await subscribe(1, 7);
    await subscribe(1, 7);
    const rows = await countRows(db, 1, 7);
    expect(rows).toBe(1);
  });

  dbTest("unsubscribe", async ({ db }) => {
    const { subscribe, unsubscribe } = await import(".");

    await insertTestUser(db);
    await insertTestPodcasts(db);

    await subscribe(1, 9);
    const rowsAfterSubscribe = await countRows(db, 1, 9);
    expect(rowsAfterSubscribe).toBe(1);

    await unsubscribe(1, 9);
    const rowsAfterUnsubscribe = await countRows(db, 1, 9);
    expect(rowsAfterUnsubscribe).toBe(0);
  });

  dbTest("subscribed podcasts", async ({ db }) => {
    const { subscribe, getSubscribedPodcasts } = await import(".");

    await insertTestUser(db);
    await insertTestPodcasts(db);
    await subscribe(1, 9);
    await subscribe(1, 1);
    await subscribe(1, 12);
    await subscribe(1, 4);

    const p = await getSubscribedPodcasts(1);
    expect(p).toMatchSnapshot();
  });
});
