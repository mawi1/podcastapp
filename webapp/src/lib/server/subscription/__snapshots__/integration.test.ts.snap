// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`playlist > subscribed podcasts 1`] = `
[
  {
    "id": 9,
    "latestEpisode": 2024-04-06T21:00:00.072Z,
    "title": "8-Bit Adventures",
  },
  {
    "id": 12,
    "latestEpisode": 2024-04-13T06:00:00.096Z,
    "title": "+Positive Vibes",
  },
  {
    "id": 4,
    "latestEpisode": 2024-02-21T13:00:00.032Z,
    "title": "The Paranormal Pundits",
  },
  {
    "id": 1,
    "latestEpisode": 2024-02-25T02:00:00.008Z,
    "title": "The Startup Squad",
  },
]
`;
